File: algorithm/sort/insertion_sort_2.c
void insertion_sort_2(int *arr, const int size) {
    for (int i = 1; i < size; i++) {
        int idx = i;
        while (idx > 0 && arr[idx - 1] > arr[idx]) {
            const int temp = arr[idx];
            arr[idx] = arr[idx - 1];
            arr[idx - 1] = temp;
            idx -= 1;
        }
    }
}

File: algorithm/sort/InsertionSort2.java
package com.example;

import java.util.Arrays;

public class InsertionSort2 {
    /**
     * 원본 배열은 수정하지 않고 새로운 배열 반환
     */
    public int[] sort(int[] arr) {
        int[] copiedArr = Arrays.copyOf(arr, arr.length);
        for (int i = 1; i < copiedArr.length; i++) {
            int idx = i;
            int originNum = copiedArr[i];
            while (idx > 0 && copiedArr[idx - 1] > originNum) {
                copiedArr[idx] = copiedArr[idx - 1];
                idx -= 1;
            }
            copiedArr[idx] = originNum;
        }
        return copiedArr;
    }
}
